### PR TITLE
Improved `neo4j-admin help` output.

### PR DESCRIPTION
**Pull request by srbaker**
### General help

```
$ bin/neo4j-admin help
Usage:

neo4j-admin set-initial-password <password>

    Sets the initial password of the initial admin user ('neo4j').

neo4j-admin dump --database=<database> --to=<destination-path>

    Dump a database into a single-file archive. The archive can be used by the load
    command. <destination-path> can be a file or directory (in which case a file
    called <database>.dump will be created). It is not possible to dump a database
    that is mounted in a running Neo4j server.

neo4j-admin load --from=<archive-path> --database=<database> [--force]

    Load a database from an archive. <archive-path> must be an archive created with
    the dump command. <database> is the name of the database to create. Existing
    databases can be replaced by specifying --force. It is not possible to replace a
    database that is mounted in a running Neo4j server.

neo4j-admin check-consistency --database=<database> [--additional-config=<file>] [--verbose]

    Check the consistency of a database.

neo4j-admin import --database=<database-name> [--mode={csv|database}] [--additional-config=<config-file-path>] [--from=<source-directory>] [--report-file=<filename>] [--nodes[:Label1:Label2]="<file1>,<file2>,..."] [--relationships[:RELATIONSHIP_TYPE]="<file1>,<file2>,..."] [--id-type=<id-type>] [--input-encoding=<character-set>] [--page-size=<page-size>]

    Import a collection of CSV files with --mode=csv (default), or a database from
    a pre-3.0 installation with --mode=database.

    --database=<database-name>
            The name of the new database to import into. This cannot be an existing
            database.
    --additional-config=<config-file-path>
            Configuration file to supply additional configuration values to
            the import tools.

    --mode=database
            Import a database from a pre-3.0 Neo4j installation.
    --from=<source-directory>
            <source-directory> is the location of the pre-3.0 database
            (e.g. <neo4j-root>/data/graph.db).

    --mode=csv
            Import a database from a collection of CSV files.
    --report-file=<filename>
            File name in which to store the report of the import.
            Defaults to import.report in the current directory.
    --nodes[:Label1:Label2]="<file1>,<file2>,..."
            Node CSV header and data. Multiple files will be logically seen as
            one big file from the perspective of the importer. The first line
            must contain the header. Multiple data sources like these can be
            specified in one import, where each data source has its own header.
            Note that file groups must be enclosed in quotation marks.
    --relationships[:RELATIONSHIP_TYPE]="<file1>,<file2>,..."
            Relationship CSV header and data. Multiple files will be logically
            seen as one big file from the perspective of the importer. The first
            line must contain the header. Multiple data sources like these can be
            specified in one import, where each data source has its own header.
            Note that file groups must be enclosed in quotation marks.
    --id-type=<id-type>
            Each node must provide a unique id. This is used to find the correct
            nodes when creating relationships. Must be one of:
                STRING: (default) arbitrary strings for identifying nodes.
                INTEGER: arbitrary integer values for identifying nodes.
                ACTUAL: (advanced) actual node ids. The default option is STRING.
            For more information on id handling, please see the Neo4j Manual:
            http://neo4j.com/docs/operations-manual/current/deployment/#import-tool
    --input-encoding <character-set>
            Character set that input data is encoded in. Defaults to UTF-8.

neo4j-admin help

    Display this help text.
```
### When a command is used incorrectly

```
$ bin/neo4j-admin check-consistency
neo4j-admin check-consistency --database=<database> [--additional-config=<file>] [--verbose]

    Check the consistency of a database.

Missing argument 'database'
```
### General help

```
$ bin/neo4j-admin help
Usage: neo4j-admin <command>

Available commands:
    set-initial-password
        Sets the initial password of the initial admin user ('neo4j').
    dump
        Dump a database into a single-file archive.
    load
        Load a database from an archive created with the dump command.
    check-consistency
        Check the consistency of a database.
    import
        Import from a collection of CSV files or a pre-3.0 database.
    help
        This help text, or help for the command specified in <command>.

Use neo4j-admin help <command> for more details.
```
### When a command is used incorrectly

```
$ bin/neo4j-admin check-consistency blah
Missing argument 'database'
neo4j-admin check-consistency --database=<database-name> [--additional-config=<config-file-path>] [--verbose]

Check the consistency of a database.

--database=<database>
        The name of the database to check the consistency of.
--additional-config=<config-file-path>
        Configuration file to supply additional configuration in.

```
### Help on a specific command

Import has a _lot_ of options, so it's the best example of where improvements might be made.

```
$ bin/neo4j-admin help import
neo4j-admin import --database=<database-name> [--mode={csv|database}] [--additional-config=<config-file-path>] [--from=<source-directory>] [--report-file=<filename>] [--nodes[:Label1:Label2]="<file1>,<file2>,..."] [--relationships[:RELATIONSHIP_TYPE]="<file1>,<file2>,..."] [--id-type=<id-type>] [--input-encoding=<character-set>] [--page-size=<page-size>]

Import a collection of CSV files with --mode=csv (default), or a database from
a pre-3.0 installation with --mode=database.

--database=<database-name>
        The name of the new database to import into. This cannot be an existing
        database.
--additional-config=<config-file-path>
        Configuration file to supply additional configuration values to
        the import tools.

--mode=database
        Import a database from a pre-3.0 Neo4j installation.
--from=<source-directory>
        <source-directory> is the location of the pre-3.0 database
        (e.g. <neo4j-root>/data/graph.db).

--mode=csv
        Import a database from a collection of CSV files.
--report-file=<filename>
        File name in which to store the report of the import.
        Defaults to import.report in the current directory.
--nodes[:Label1:Label2]="<file1>,<file2>,..."
        Node CSV header and data. Multiple files will be logically seen as
        one big file from the perspective of the importer. The first line
        must contain the header. Multiple data sources like these can be
        specified in one import, where each data source has its own header.
        Note that file groups must be enclosed in quotation marks.
--relationships[:RELATIONSHIP_TYPE]="<file1>,<file2>,..."
        Relationship CSV header and data. Multiple files will be logically
        seen as one big file from the perspective of the importer. The first
        line must contain the header. Multiple data sources like these can be
        specified in one import, where each data source has its own header.
        Note that file groups must be enclosed in quotation marks.
--id-type=<id-type>
        Each node must provide a unique id. This is used to find the correct
        nodes when creating relationships. Must be one of:
            STRING: (default) arbitrary strings for identifying nodes.
            INTEGER: arbitrary integer values for identifying nodes.
            ACTUAL: (advanced) actual node ids. The default option is STRING.
        For more information on id handling, please see the Neo4j Manual:
        http://neo4j.com/docs/operations-manual/current/deployment/#import-tool
--input-encoding <character-set>
        Character set that input data is encoded in. Defaults to UTF-8.
```
### Help on a command that doesn't exist

```
$ bin/neo4j-admin help foobar
Unknown command: foobar. Available commands are: set-initial-password dump load check-consistency import

neo4j-admin help [<command>]

This help text, or help for the command specified in <command>.
```

### CODE REVIEW

File: community/command-line/src/main/java/org/neo4j/commandline/admin/AdminTool.java [hunk 1]
```diff
@@ -104,9 +104,9 @@ public void execute( Path homeDir, Path configDir, String... args )
 
     private void badUsage( AdminCommand.Provider command, IncorrectUsage e )
     {
+        failure( e.getMessage() );
```
Comment (benbc) [inline community/command-line/src/main/java/org/neo4j/commandline/admin/AdminTool.java]: This is not going to work -- `failure()` calls `System.exit()`.

There is also another version of `badUsage()` where the order needs to change.

File: community/command-line/src/test/java/org/neo4j/commandline/admin/HelpCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    @Mock
+    private Consumer<String> out;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Ignore
+    public void showsGeneralHelpOutputWhenNoCommandIsProvided() {};
```
Comment (benbc) [inline community/command-line/src/test/java/org/neo4j/commandline/admin/HelpCommandTest.java]: Bong!

File: community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java [hunk 3]
```diff
@@ -63,9 +72,34 @@ public HelpCommand( Usage usage, Consumer<String> output )
...
+        {
+            usage.print( output );
+        }
+    }
+
     @Override
     public void execute( String... args )
     {
-        usage.print( output );
+        executeWithCommandLocator( CommandLocator.fromServiceLocator(), args );
```
Comment (benbc) [inline community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java]: Can't you inject the `CommandLocator` in the constructor?

File: community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java [hunk 4]
```diff
@@ -48,6 +51,12 @@ public String description()
...
         @Override
+        public String summary()
+        {
+            return description();
+        }
+
+        @Override
         public AdminCommand create( Path homeDir, Path configDir, OutsideWorld outsideWorld )
         {
             return new HelpCommand( usage, outsideWorld::stdOutLine );
```
Comment (benbc) [inline community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java]: Is the `usage` parameter still used?

**Review by benbc — CHANGES_REQUESTED** [review]

File: community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java (hunk 4 above)
Comment (srbaker) [reply community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java]: Yes, we use it for the default case of simply `neo4j-admin help`

File: community/command-line/src/main/java/org/neo4j/commandline/admin/AdminTool.java [hunk 5]
```diff
@@ -105,14 +105,16 @@ public void execute( Path homeDir, Path configDir, String... args )
...
         commandUsage.print( outsideWorld::stdErrLine );
-        failure( e.getMessage() );
+        failure();
     }
 
     private void badUsage( String message )
     {
         usage.print( outsideWorld::stdErrLine );
-        failure( message );
+        outsideWorld.stdErrLine( message );
```
Comment (benbc) [inline community/command-line/src/main/java/org/neo4j/commandline/admin/AdminTool.java]: But shouldn't this now go above the usage?

File: community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java [hunk 6]
```diff
@@ -38,34 +41,62 @@ public Provider( Usage usage )
...
+                Usage.CommandUsage commandUsage = new Usage.CommandUsage( commandProvider, "neo4j-admin" );
+                commandUsage.print( this.output );
+                this.output.accept( commandProvider.description() );
+                this.output.accept( "" );
+                return;
+            }
+            catch (NoSuchElementException e)
+            {
+                this.output.accept( "Unknown command: " + args[0] );
+                this.output.accept( "" );
```
Comment (benbc) [inline community/command-line/src/main/java/org/neo4j/commandline/admin/HelpCommand.java]: This case should exit non-zero, I think. That is, it needs to throw an exception. How about also arranging for it to print the valid list of commands?

File: community/command-line/src/test/java/org/neo4j/commandline/admin/HelpCommandTest.java (hunk 2 above)
Comment (benbc) [reply community/command-line/src/test/java/org/neo4j/commandline/admin/HelpCommandTest.java]: This needs to go or be implemented.

**Review by benbc — COMMENTED** [review]
Please can you show before (maybe truncted) and after output in the PR description?

**Review by benbc — APPROVED** [review]

**Comment by benbc** [comment]
LGTM

**Pull request merged**